runelite-client/src/main/java/net/runelite/client/plugins/clanchat/ClanChatPlugin.java

```diff
@@ -70,22 +97,132 @@ public void updateClanChatTitle()
 		}
 	}
 
+	/**
+	 * Takes incoming cc messages, checks for rank, adds rank imgage,
+	 * and filters Total Levels below if active. Stores previously filtered
+	 * accounts in a hashSet to avoid making redudant hiscore lookups.
+	 *
+	 * @param setMessage message to be altered
+	 */
 	@Subscribe
 	public void onSetMessage(SetMessage setMessage)
 	{
+
+	    boolean ranked;
+	    //If filter minimum has changed, flush the stored accounts
+	    if (currentSetFilter != config.getTotalLevelFilter())
+		{
+			filteredAccounts.clear();
+			currentSetFilter = config.getTotalLevelFilter();
+		}
+
 		if (client.getGameState() != GameState.LOADING && client.getGameState() != GameState.LOGGED_IN)
 		{
 			return;
 		}
 
 		if (setMessage.getType() == ChatMessageType.CLANCHAT && client.getClanChatCount() > 0)
 		{
-			insertClanRankIcon(setMessage);
+			//Look through hash for acct if already to be filtered
+			if (isPlayerFiltered(setMessage.getName()))
+			{
+				log.debug("player: " +setMessage.getName() + " is already filtered");
+				filterMessage(setMessage);
+			}
+			//lookup rank and hiscore lookup if TL filter is active and account is not ranked in cc
+			else
+			{
+				ranked = insertClanRankIcon(setMessage);	//Check for ranked
+				if (config.getTotalLevelFilter() != 0 && !ranked)
+				{
+					executor.submit(() -> lookupPlayer(setMessage));
+					return;
+				}
+			}
+			client.refreshChat();
+		}
+	}
+
+	/**
+	 * Looks up player on hiscores and alters the message if it is to be filtered.
+	 * @param message account name pulled from message and altered if neede
+	 */
+	private void lookupPlayer(final SetMessage message)
+	{
+		String account = message.getMessageNode().getName();
+
+		account = sanitize(account);
+
+		HiscoreSkill skill;
+		skill =  HiscoreSkill.valueOf(net.runelite.api.Skill.OVERALL.getName().toUpperCase());
+
+		if (Strings.isNullOrEmpty(account))
+		{
+			return;
+		}
+		try
+		{
+			SingleHiscoreSkillResult result = hiscoreClient.lookup(account, skill);
+			Skill hiscoreSkill = result.getSkill();
+
+			if (config.getTotalLevelFilter() > hiscoreSkill.getLevel())
+			{
+				//Add account to "blacklist" hashSet
+				filteredAccounts.add(message.getMessageNode().getName());
```

Comment:
Need to take care of concurrency. `filteredAccounts` need to be thread safe data structure

Reply:
Going to take a look into ConncurrentHashMap for thread safe data as well as the ability to store TL along with Account Name as recommended above. Will update once I make some headway.